Swift: Extract typealias relations

geoffw0 opened the pull request:
Extract the aliased type for a `typealias`, for example `B` and `Int?` in:
```
typealias A = B
typealias C = Int?
```
Thanks to @MathiasVP for getting me started on this.  In addition to the extractor change itself there is:
- QLDoc (including generated QLDoc).
- tests (extractor and library).
- a new helper method `Type.getUnderlyingType()` (similar to in CPP).
- a small but important fix for `NominalType.getABaseType()`.
- upgrade and downgrade scripts.
  - I've tested the upgrade script but I could not figure out how to test the downgrade script.
  - Supposedly there is "some automated testing of the scripts" anyway.
  - If there's anything else I can do to check this is right please ask.

geoffw0 commented:
It looks like there's a problem related to some built-in type aliases?
```
+[VALUE_NOT_IN_TYPE] predicate any_generic_types(@any_generic_type id, @decl_or_none declaration): Value 113759 of field declaration is not in type @decl_or_none. Appears in tuple (113758,113759)
+	Relevant element: declaration=113759
+		Full ID for 113759: @"StructDecl_(112489)So13__va_list_tagV". The ID may expand to @"StructDecl_{@"ModuleDecl___ObjC"}So13__va_list_tagV"
+[VALUE_NOT_IN_TYPE] predicate type_alias_types(@type_alias_type id, @type_alias_decl_or_none decl): Value 6073 of field decl is not in type @type_alias_decl_or_none. Appears in tuple (6072,6073)
+	Relevant element: decl=6073
+		Full ID for 6073: @"TypeAliasDecl_(65)$sSo9AnyObjectyXlmF". The ID may expand to @"TypeAliasDecl_{@"ModuleDecl_Builtin"}$sSo9AnyObjectyXlmF"
+[VALUE_NOT_IN_TYPE] predicate type_alias_types(@type_alias_type id, @type_alias_decl_or_none decl): Value 38308 of field decl is not in type @type_alias_decl_or_none. Appears in tuple (38306,38308)
+	Relevant element: decl=38308
+		Full ID for 38308: @"TypeAliasDecl_(65)$sSo17UnsafeValueBufferBBmF". The ID may expand to @"TypeAliasDecl_{@"ModuleDecl_Builtin"}$sSo17UnsafeValueBufferBBmF"
+[VALUE_NOT_IN_TYPE] predicate type_alias_types(@type_alias_type id, @type_alias_decl_or_none decl): Value 113319 of field decl is not in type @type_alias_decl_or_none. Appears in tuple (113757,113319)
+	Relevant element: decl=113319
+		Full ID for 113319: @"TypeAliasDecl_(112489)$sSo17__builtin_va_listSo02__b1_C4_tagVmF". The ID may expand to @"TypeAliasDecl_{@"ModuleDecl___ObjC"}$sSo17__builtin_va_listSo02__b1_C4_tagVmF"
```
I'll have to think about this next week.

d10c commented:
You will need to add the builtin symbols `UnsafeValueBuffer` and `AnyObject` to [swift/extractor/SwiftBuiltinSymbols.h](https://github.com/github/codeql/blob/241951f53e094100f11e44ace6847f1dd7fba5d3/swift/extractor/SwiftBuiltinSymbols.h). However, the `Builtin` module is not the only one that needs to be extended with new builtin symbols; you'll need to do the same thing for the `_ObjC` module (symbols `__va_list_tag` and `__builtin_va_list`).

geoffw0 commented:
> you'll need to do the same thing for the _ObjC module (symbols __va_list_tag and __builtin_va_list).

Am I right in understanding that I need to create a new mechanism to do this, much like `swiftBuiltins` but for a module named `_ObjC`, rather than for one where `isBuiltinModule` holds.

I gather no upgrade / downgrade script changes are required for this, as the builtins are just `Decl`s in the existing schema.

geoffw0 commented:
I've pushed an attempt at fixing this, we'll see what the `qltests-linux` PR check thinks of it.

Note that the upgrade / downgrade scripts will need updating before this can be merged, but I don't want to do that prematurely.

d10c commented on swift/extractor/SwiftBuiltinSymbols.h:
```diff
@@ -69,5 +70,12 @@ constexpr std::array swiftBuiltins = {
+    "AnyObject",
+};
+
+std::vector<std::string> objcBuiltins {
```
Hmm, apparently the DB-CHECKs are still failing for these new builtins. 

Btw, is there a reason you went for std::vector instead of a constexpr array? This is still a header, so static and/or constexpr sounds more appropriate.

AlexDenisov replied:
> Hmm, apparently the DB-CHECKs are still failing for these new builtins.

jfyi, I'm looking into it now

geoffw0 replied:
> is there a reason you went for std::vector instead of a constexpr array?

Because the two arrays would actually have different template types (`std::array<T, N>` where T = `std::string` or `const char *` or whatever in both cases, but `N` = the length of the array, which differs).  That makes passing it to a function a pain as the function would have to be a template on `N` (unless I'm missing a trick).

d10c replied:
I believe you can monomorphize various types of arrays into a std::span?

geoffw0 commented:
I've just force pushed, removed three commits - two that attempted to fix the extraction issue myself, one that was an old merge with main.  Then I've added a new merge with main which should include the new fix by @AlexDenisov .  This PR is still not fit to merge until I repair the upgrade/downgrade scripts, but I want to see the tests pass first.

geoffw0 commented:
... well the qltests we were looking at have passed!  Something else has gone wrong though ...

geoffw0 commented:
All tests pass except for the upgrade script.  I shall start a DCA run for peace of mind, then see about repairing the upgrade script.

geoffw0 commented:
I'm struggling to get the upgrade / downgrade scripts working even though I have old code for them that works.  I think too much _stuff_ has accumulated on this PR making it difficult to manage.  Closing.  I'll reconstruct everything in smaller steps.

geoffw0 commented:
First round of replacement PRs:
 - https://github.com/github/codeql/pull/12485
 - https://github.com/github/codeql/pull/12489

geoffw0 commented:
Main PR:
- https://github.com/github/codeql/pull/12509

geoffw0 commented:
Another PR that came from this:
- https://github.com/github/codeql/pull/12520
